packages/audioplayers/tizen/src/audio_player.cc

```diff
@@ -208,138 +189,120 @@ int AudioPlayer::GetDuration() {
+  ret = player_set_looping(player_, (release_mode_ == ReleaseMode::kLoop));
+  HandleResult("player_set_looping", ret);
 
-  LOG_DEBUG("prepare audio player asynchronously");
-  result = player_prepare_async(player_, OnPrepared, (void *)this);
-  HandleResult("player_prepare_async", result);
+  ret = player_prepare_async(player_, OnPrepared, (void *)this);
+  HandleResult("player_prepare_async", ret);
   preparing_ = true;
   seeking_ = false;
 }
 
+void AudioPlayer::EmitPositionUpdates() {
+  ecore_main_loop_thread_safe_call_async(StartPositionUpdates, (void *)this);
+}
+
 void AudioPlayer::ResetPlayer() {
-  LOG_INFO("reset audio player...");
-  int result;
+  int ret;
   player_state_e state = GetPlayerState();
   switch (state) {
     case PLAYER_STATE_NONE:
       CreatePlayer();
       break;
     case PLAYER_STATE_IDLE:
       if (preparing_) {
-        LOG_DEBUG("player is preparing, unprepare the player");
-        result = player_unprepare(player_);
-        HandleResult("player_unprepare", result);
+        // Cancel preparing if it's already preparing.
+        ret = player_unprepare(player_);
+        HandleResult("player_unprepare", ret);
         preparing_ = false;
       }
       break;
     case PLAYER_STATE_READY:
     case PLAYER_STATE_PLAYING:
     case PLAYER_STATE_PAUSED:
-      LOG_DEBUG("unprepare audio player");
-      result = player_unprepare(player_);
-      HandleResult("player_unprepare", result);
+      ret = player_unprepare(player_);
+      HandleResult("player_unprepare", ret);
       break;
   }
 }
 
 player_state_e AudioPlayer::GetPlayerState() {
-  player_state_e state;
-  if (player_ == nullptr) {
-    state = PLAYER_STATE_NONE;
-  } else {
-    int result = player_get_state(player_, &state);
-    if (result != PLAYER_ERROR_NONE) {
-      state = PLAYER_STATE_NONE;
+  player_state_e state = PLAYER_STATE_NONE;
+  if (player_) {
+    int ret = player_get_state(player_, &state);
+    if (ret != PLAYER_ERROR_NONE) {
+      LOG_ERROR("Getting player(id: %s) state failed: %s\n", player_id_.c_str(),
+                get_error_message(ret));
     }
   }
   return state;
 }
 
-void AudioPlayer::HandleResult(const std::string &func_name, int result) {
-  if (result != PLAYER_ERROR_NONE) {
-    std::string error(get_error_message(result));
-    LOG_ERROR("%s failed : %s", func_name.c_str(), error.c_str());
-    throw AudioPlayerError(error, func_name + " failed");
-  }
-}
-
 void AudioPlayer::OnPrepared(void *data) {
-  LOG_INFO("Audio player is prepared");
   AudioPlayer *player = (AudioPlayer *)data;
   player->preparing_ = false;
 
   int duration = 0;
-  int result = player_get_duration(player->player_, &duration);
-  if (result == PLAYER_ERROR_NONE) {
+  int ret = player_get_duration(player->player_, &duration);
+  if (ret == PLAYER_ERROR_NONE) {
     player->prepared_listener_(player->player_id_, duration);
   }
 
   player_set_playback_rate(player->player_, player->playback_rate_);
```

Comment:
It seems error handling was missing in the original code, I've fixed the code so that it throws `AudioPlayerError` on error.

Reply:
Uhm, that's not good. Who will catch the exception thrown in the callback?

Reply:
Oh I see, static callback functions should pass errors with `error_listener_` as exceptions thrown in callbacks will not be caught by try-catch block in `HandleMethodCall`.

Reply:
@swift-kim 
How about changes in https://github.com/flutter-tizen/plugins/pull/357/commits/4014953fdd17fce43a3bfc3869a5b2450dd9ab3d?

Basically the code defines an explicit return type `Result` (similar to geolocator's `TizenResult`), but it also defines another type `ResultOr<T>` to also return values. 

I've reference the following resources:
- [error-handling-with-statusor](https://www.furidamu.org/blog/2017/01/28/error-handling-with-statusor/)
- [protobuf](https://github.com/protocolbuffers/protobuf/blob/main/src/google/protobuf/stubs/statusor.h)

I still left some exception mechanism in `HandleMethodCall` because it significantly removes nested if clauses and code duplication.

Reply:
Well, please be reluctant to inventing something surprising or unusual which may not be easily understood by average readers. The goal of this refactoring project is to reduce such things by defining a common rule that applies to a wide range of packages.

Reply:
Okay, I reference geolocator's [`TizenResult`](https://github.com/flutter-tizen/plugins/blob/master/packages/geolocator/tizen/src/tizen_result.h) and used [output parameters](https://google.github.io/styleguide/cppguide.html#Inputs_and_Outputs).